exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/unnest/UnnestImpl.java

```diff
@@ -48,24 +51,47 @@
   private static final Logger logger = LoggerFactory.getLogger(UnnestImpl.class);
 
   private ImmutableList<TransferPair> transfers;
-  private LateralContract lateral; // corresponding lateral Join (or other operator implementing the Lateral Contract)
   private SelectionVectorMode svMode;
   private RepeatedValueVector fieldToUnnest;
   private RepeatedValueVector.RepeatedAccessor accessor;
   private RecordBatch outgoing;
 
+  private IntVector rowIdVector ; // Allocated and owned by the UnnestRecordBatch
+  private IntVector.Mutator rowIdVectorMutator;
+
   /**
    * The output batch limit starts at OUTPUT_ROW_COUNT, but may be decreased
    * if records are found to be large.
    */
   private int outputLimit = ValueVector.MAX_ROW_COUNT;
 
 
-  // The index in the unnest column that is being processed.We start at zero and continue until
+
+  /**
+   * We maintain three indexes
+   *
+   *
+   *
+                valueIndex  0         1       2   3
+                            |- - - - -|- - - -|- -|- - - -|
+                            | | | | | | | | | | | | | | | |
+                            |- - - - -|- - - -|- -|- - - -|
+           innerValueIndex  0 1 2 3 4 0 1 2 3 0 1 0 1 2 3 |
+     runningInnerValueIndex 0 1 2 3 4 5 6 7 8 9 ...
+   *
+   *
+   *
+   */
+  private int valueIndex; // index in the incoming record being processed
+  // The index in the unnest column that is being processed. We start at zero and continue until
   // InnerValueCount is reached or  if the batch limit is reached
   // this allows for groups to be written between batches if we run out of space, for cases where we have finished
   // a batch on the boundary it will be set to 0
```

Comment:
please revisit this comment, it's little confusing. How about just saying _"It's the index of the array element in the unnest column at row pointed by valueIndex which is currently being processed. It starts at zero and continue until InnerValueCount is reached or the batch limit is reached. It allows for groups to be written across batches if we run out of space"_